### ISSUE TITLE
About dense_to_sparse

### ISSUE THREAD

**Issue by shawnwang-tech** [rusty1s/pytorch_geometric#414]
## ❓ Questions & Help

I want to use `dense_to_sparse` in my code. I am wondering if it is derivable (the grad can be passed through it)?

Thanks in advanced!

**Comment by rusty1s** [comment]
Sure, but only for non-zero entries :)

**Comment by shawnwang-tech** [comment]
This may be the reason that my code does not work as my expect. Maybe I have to compute the zero entries' gradient? I am using a Gumbel-softmax to generate graph structure (a dense N*N matrix), and use `dense_to_sparse` to generate `edge_index` for GNN. But I found that Gumbel Leaner can learn nothing. Because of the lack of the zero entries' gradient? Do you have any suggestions? Thank you!

**Comment by rusty1s** [comment]
Well, you can't compute the zero gradients because there are none if you proceed with a sparse graph. This does not have to mean anything bad by default, *e.g.*, ReLU or topk do remove information/gradients all the time.

You should check the gradients of your model more closely to find out where those may get lost.

**Comment by shawnwang-tech** [comment]
Thank you for your help!!!! I will read the paper and check my model.

**Comment by shawnwang-tech** [comment]
I followed your suggestion and checked my code closely. Then I found where the gradients get lost. In the Gumbel training process, Gumbel generates a dense adjacency-like matrix in which each element is a float number range from 0 to 1, not just 0 or 1, I want to fork `pytorch_scatter`, and pass `value` along with `index` to `scatter_add`. Make `edge_index` be 'dense' (contains every edge with the value indicates the connection probability). make `scatter_add` to be `scatter_weighted_add`. Is this possible? Thanks!

**Comment by rusty1s** [comment]
Hi, I do not understand what you want to scatter. Can you elaborate?

**Comment by shawnwang-tech** [comment]
<img width="418" alt="image" src="https://user-images.githubusercontent.com/16516703/59658118-4e589300-91d5-11e9-9e76-883959070037.png">
 add weights here.

**Comment by rusty1s** [comment]
You can pre-multiply weights to the input elements.

**Comment by shawnwang-tech** [comment]
Oh, yes. That may solve my problem. Thank you, I will try!

**Comment by shawnwang-tech** [comment]
Problem solved using pre-multiplying edge_value to input! Thanks for all your help!!!

**Comment by rusty1s** [comment]
If you convert from dense to sparse, you have to incorporate the edge weights. Otherwise, this is a discrete operation that will lose gradients.
